src/extension.ts

```diff
@@ -8,90 +8,13 @@ import ActivityLog from "./activityLog";
 const EXTENSION_NAME = "liveshare";
 const POLICY_FILE = "liveshare-policy.json";
 
-const SETTINGS = [
-  "allowGuestDebugControl",
-  "allowGuestTaskControl",
-  "autoShareServers",
-];
-
-async function enforceSettings() {
-  const config = vscode.workspace.getConfiguration(EXTENSION_NAME);
-  return Promise.all(
-    SETTINGS.map((setting) =>
-      config.update(setting, false, vscode.ConfigurationTarget.Global)
-    )
-  );
-}
-
 export async function activate(context: vscode.ExtensionContext) {
-  // Ensure the necessary settings are
-  // re-enforced for the end-user.
-  await enforceSettings();
-
   // This extension takes a hard dependency on
   // Live Share, so it will always be available.
   const api = (await vsls.getApi())!;
 
-  // Wait to see if the host attempts to share,
-  // and is using a disallowed identity.
-  api.onDidChangeSession(async (e) => {
-    if (e.session) {
-      const allowedDomains = getAllowedDomains();
-
-      if (
-        allowedDomains.length > 0 &&
-        !allowedDomains.includes(getDomain(api.session)!)
-      ) {
-        const domains = allowedDomains.sort().join(", ");
-        api.end();
-
-        if (
-          await vscode.window.showErrorMessage(
-            `You need to sign into Live Share using one of the following allowed domains: ${domains}. Please sign-in and share again.`,
-            "Sign in"
-          )
-        ) {
-          await reSignIn();
-        }
-      }
-    }
-  });
-
-  // Wait for any guests to attempt to join
-  // a collaboration session, in order to
-  // determine if they're allowed in or not.
-  api.onDidChangePeers((e) => {
-    e.added.forEach((peer) => {
-      // If the current user doesn't have an e-mail address, then
-      // there's no point in us trying to match it against guests.
-      const selfDomain = getDomain(api.session);
-      if (!selfDomain) {
-        return;
-      }
-
-      // If the incoming user doesn't have an e-mail address,
-      // then immediately remove them, since anonymous/unknown
-      // guests aren't allowed, regardless of the host's settings.
-      const emailDomain = getDomain(peer);
-      if (!emailDomain) {
-        return removeUser(peer.peerNumber);
-      }
-
-      // If the incoming user is from the same domain
-      // as the host, then they're immediately allowed.
-      if (selfDomain.localeCompare(emailDomain) === 0) {
-        return;
-      }
-
-      const allowedDomains = getAllowedDomains();
-
-      // If the incoming user is from a different domain,
-      // which hasn't been whitelisted, then remove them.
-      if (!allowedDomains.includes(emailDomain)) {
-        removeUser(peer.peerNumber);
-      }
-    });
-  });
+  let policyProvider = new GatekeeperPolicyProvider();
+  api.registerPolicyProvider("Gatekeeper Provider", policyProvider);
```

Comment:
Just to confirm: this method returns a disposeable? That when disposed, would remove the old policies from Live Share? And then we could re-register the policy provider again to "refresh" the settings?

Reply:
Yep, I showcased this scenario in my latest commit.